### PR TITLE
Use "gender" as a stopword only when it's visible

### LINKED ISSUE #49: "Gender" should not act as a stopword if it appears only in text not visible on the screen
Since the href of a link is not visible on the screen, this is a confusing behavior.

### PR DESCRIPTION
**Pull request by beszel**
Resolves #49.
@ProfJanetDavis If I should be handling the commit history differently – for example, if I should be squashing commits or something – please let me know!
Another area that may need attention is whether or not the tests have been implemented in the correct files.

### CODE REVIEW

File: src/stopword-highlights.js [hunk 1]
```diff
@@ -17,6 +18,20 @@ export function mentionsGender(text) {
     return text.match(/gender/i) ? true : false;
 }
 
+// Report whether the page includes the keyword "gender" visibly.
+export function visiblyMentionsGender(body) {
+    const textNodes = textNodesUnder(body);
+    let node = null;
+    for (node of textNodes) {
+        if (mentionsGender(node.data) && isVisible(node)) {
+            // If any text node mentions gender,
```
Comment (ProfJanetDavis) [inline src/stopword-highlights.js]: I haven't seen this style of commenting on conditionals before. I like it.

File: src/stopword-highlights.js [hunk 2]
```diff
@@ -17,6 +18,20 @@ export function mentionsGender(text) {
...
+            return true;
+        }
+    }
+    // If no text nodes mention gender,
+    return false;
```
Comment (ProfJanetDavis) [inline src/stopword-highlights.js:32]: I haven't seen this style of commenting on conditionals before. I like it.

File: test/e2e/content/case.nothing-to-change.test.js [hunk 3]
```diff
@@ -33,4 +36,9 @@ describe("When the page does not include any gender pronouns or stopwords, it",
         const headerText = await page.$eval(headerSelector, e => e.innerText);
         expect(headerText).to.include("no gender pronouns");
     });
+
+    it("should make replacements even when hidden elements contain the word 'gender'", async function() {
+        const replacements = visibilityPage.$$(replacementSelector);
+        expect(replacements).to.be.empty;
+    });
 });
```
Comment (ProfJanetDavis) [inline test/e2e/content/case.nothing-to-change.test.js]: This is confusing. The description is "it should make replacements", but the test is "expect(replacements).to.be.empty". Please clarify.

Comment (ProfJanetDavis) [reply test/e2e/content/case.nothing-to-change.test.js]: Is a second test needed where pronouns are actually replaced? I'm fairly convinced this one test is adequate to ensure the right code path is being taken, but you should be convinced as well.

If you do think a second test is needed, that one would go in `test/e2e/content/case.with-gender-pronouns.test.js`

Comment (ProfJanetDavis) [reply test/e2e/content/case.nothing-to-change.test.js]: If a single test, it is probably more appropriate to put it in a new file named something like `case.mentions-gender-invisibly.js`. The new test you wrote has no code in common with the existing tests in `case.nothing-to-change.test.js`, and I would argue it is actually a new use case. I don't mind proliferating files if it keeps the intention behind each file clearer.

**Review by ProfJanetDavis — CHANGES_REQUESTED** [review]
All my comments requiring changes are on the end-to-end test. You were right to be concerned about where you put things - I don't think the test is in the right place. It also needs clarification. You'll see my thought process in the comments.

There were a few things I decided not to comment on, where a moment's thought convinced me you made the right decision.

**Comment by ProfJanetDavis** [comment]
Also - yes, please do squash these commits! I'd like to see the tests and new code together in one commit. Ideally, you'd have written the tests before the code (see https://www.agilealliance.org/glossary/tdd/), though of course that wasn't possible here as I was still building out the test architecture.

File: test/e2e/content/case.nothing-to-change.test.js (hunk 3 above)
Comment (ProfJanetDavis) [reply test/e2e/content/case.nothing-to-change.test.js]: Or, perhaps better, all of the tests in his case could use the new page. The tests that were originally there would be equally applicable to the new page you created. You might rename it accordingly.

**Comment by beszel** [comment]
What do you think of including the link and hidden elements in `helper.html`? This ticket says that an invisible instance of the word "gender" should do nothing, so we might want to modify the page being tested rather than add new tests that mirror existing tests on a new page. Does that make sense?
One drawback of this option is that the messages shown for test failure won't say anything about hidden elements. That could make debugging more difficult.

**Comment by ProfJanetDavis** [comment]
I like the idea of not proliferating HTML pages for testing. Yes, go ahead and do that.

**Comment by beszel** [comment]
I did a force-push because I mistakenly reset some changes. I hope to avoid that in the future by being more mindful of the remote.
This last commit was easier than writing new end-to-end tests, and it should either address or make irrelevant each of the comments above. Let me know if this needs anything else!

**Review by ProfJanetDavis — APPROVED** [review]
Looks good! Agreed, this is much simpler.

**Pull request merged**